### DISCUSSION TITLE
Copy and paste nodes between Godot projects

### DISCUSSION THREAD

**Discussion by Tormentous** [godotengine/godot-proposals#10397] [Editor]
Sometimes I want to copy a node I have from an older project because I think It would be perfectly suitable for the project I'm currently working on, but I find out that you can't just copy and paste between projects and you gotta do it manually

**Comment by AThousandShips** [comment]
This would be very complicated to implement, first of all the way the data would be transferred would be a lot of work and complicated, unless we make the different Godot instances communicate with each other you'd need to somehow store all this data on the clipboard, which is complicated, for example generally the clipboard contains simple data most of the time (if you copy a file in the file manager it doesn't actually just put the file data in the clipboard, it just handles it when you paste it)

And we'd assume here that simple cases would be handled directly, like copying a node with a script attached, or a simple resource like a gradient file

But also how would you handle these situations:
* A script that extends some other custom type, for example `Foo` which extends `Bar`, would it *also* copy this type if it doesn't already exist in the target project? What if it does exist but is a completely different type?
* Other dependencies, like `preload` resources in scripts?
* Imported resources, how would it handle a node being copied that uses a mesh imported from a file at `player/assets/mesh.glb`? Would it copy that mesh and the `mesh.glb.import` file? Where would it put it?

**Reply by hsandt** [reply]
I see, I just needed a very barebone copy in my case, to extract a MWE to a new project (I couldn't repro a camera bug and wanted to copy my Camera2D node with exactly the same settings). So I'd need some sort of serialized data to copy in the clipboard and paste back in the other Godot editor.

Thinking about Serialization, I had to idea of clicking on the Inspector > Gear/Tools icon > Copy properties and then Paste properties on the other project. But I realized this is also an internal copy, that only this Godot instance knows about, not a copy to clipboard...

I understand why overwriting user clipboard would be problematic for people actively using it for other things, but maybe there could be an advanced debug command to force copying node or property serialization into the clipboard specifically for cross-instance copy.

For now, the best seems to be to copy the scene .tscn and then remove unwanted parts to extract the MWE, but surely I'll get a lot of dependency breaking warnings. So I'll have to create a scene with just the elements I need in the source project, then copy that scene into the MWE project...

**Reply by hsandt** [reply]
Funny, I just found out that when creating a new scene in 4.4.1, there is a new option "Paste From Clipboard"

![image](https://github.com/user-attachments/assets/510840ed-662b-4662-af99-fd9af33aec4e)

However it only seems to appear after copying a node from the same project, so it still uses the "internal clipboard" and not the system clipboard. It makes sense, we'd have to find a way to copy node info to system clipboard before we can paste it, after all.

**Reply by theraot** [reply]
The way I would wish/hope this feature goes about would be to save the selection to a new scene, but not to disk but to ram, and put the text resource into the clipboard. Yes, it is not a deep copy, it will not bring dependencies, etc. But if I wanted something like that, I would be copying files, not nodes from the scene dock. Then you can bring up the dialog to fix dependencies when pasted, and let me figure it out.

**Comment by Mickeon** [comment]
This _may_ be feasible in two ways I just thought of on the spot. Just spitballing here:
- Convert the copied nodes into string and insert into the system's clipboard
	- The copied nodes could be packed as a scene and the clipboard could contain the PackedScene, in `.tscn` format. 
	- The extended support for JSON objects may help, but is not enough, because the node structure needs to be copied, as well.
	- The system's clipboard filled with "noise" could potentially be unexpected and undesirable.
- Save copied nodes on disk
	- These would be saved in a temporary editor folder, in a shared format (presumably `.scn`)
	- Would the editor seriously have to check for this file every time a paste operation is executed? Sounds highly inefficient, but maybe the OS knows better

Regardless, data loss due to missing resources would be inevitable. It may even be safer to copy and paste `.tscn` files manually.

**Reply by hawkerm** [reply]
If you open a scene with missing dependencies though, there's already a dialog that exists for reselecting those with existing resources. It could be interesting to see that used for this scenario as well, so if you paste something from another editor instance it can pop that up in order to let you reparent things to your project. i.e. offload that responsibility to the developer.

Like if I copy a .tscn file to another project, this is what happens. That's all I want for copy/paste, I just want to not have to bring the whole structure over or recreate it by hand.
